pretyping/typing.ml

```diff
@@ -59,57 +59,54 @@ let assumption_of_judgment env sigma j =
   with Type_errors.TypeError _ | PretypeError _ ->
     error_assumption env sigma j
 
-let judge_of_applied_inductive_knowing_parameters env sigma funj ind argjv =
-  let rec apply_rec sigma n typ = function
-  | [] ->
-      sigma, { uj_val  = mkApp (j_val funj, Array.map j_val argjv);
-               uj_type =
-                 let ar = inductive_type_knowing_parameters env sigma ind argjv in
-                 hnf_prod_appvect env sigma (EConstr.of_constr ar) (Array.map j_val argjv) }
-  | hj::restjl ->
-       let sigma, (c1,c2) =
-         match EConstr.kind sigma (whd_all env sigma typ) with
-         | Prod (_,c1,c2) -> sigma, (c1,c2)
-         | Evar ev ->
-             let (sigma,t) = Evardefine.define_evar_as_product env sigma ev in
-             let (_,c1,c2) = destProd sigma t in
-             sigma, (c1,c2)
-         | _ ->
-             error_cant_apply_not_functional env sigma funj argjv
-       in
-       begin match Evarconv.unify_leq_delay env sigma hj.uj_type c1 with
-         | sigma ->
-           apply_rec sigma (n+1) (subst1 hj.uj_val c2) restjl
-         | exception Evarconv.UnableToUnify _ ->
-           error_cant_apply_bad_type env sigma (n, c1, hj.uj_type) funj argjv
-       end
+exception ArityMismatch
+
+let decompose_product env sigma typ args =
+  let rec decompose sigma accu subs typ = function
+  | [] -> sigma, accu, Vars.esubst Vars.lift_substituend subs typ
+  | hj :: restjl ->
+    match EConstr.kind sigma typ with
+    | Prod (_, c1, c2) ->
+      (* Fast path *)
+      let nsubs = Esubst.subs_cons (Vars.make_substituend hj.uj_val) subs in
+      decompose sigma ((hj.uj_type, subs, c1) :: accu) nsubs c2 restjl
+    | _ ->
+      let typ = Vars.esubst Vars.lift_substituend subs typ in
+      let subs = Esubst.subs_id 0 in
+      let nsubs = Esubst.subs_cons (Vars.make_substituend hj.uj_val) subs in
+      let sigma, (c1,c2) =
+        match EConstr.kind sigma (whd_all env sigma typ) with
```

Comment:
IIUC here we may be reducing an ill-typed term, and so possibly not terminate. Is that not a concern?

Reply:
Hm, this is right. I think I can partially revert the code splitting the evaluation in two phases and try to be more cautious about that. Non-termination is not the only potential issue, we might get into weird crashes / anomalies.